File: taurus-protect-sdk-go/pkg/protect/rules_cache_verification_test.go
package protect

import (
	"context"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/x509"
	"encoding/pem"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/mapper"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// The rules cache supplies the HSM public key that address signature verification
// trusts, so the container it holds must have its SuperAdmin signatures verified
// before anything reads it. An unverified container makes that check pass against
// whatever key the container carries — attacker-chosen addresses verify clean.
//
// This is a cross-SDK invariant and TypeScript had drifted off it: its cache fetched
// through the generated API and the raw mapper, skipping verification entirely. Nothing
// in any SDK tested the path, which is why it survived. Guard for this SDK.
func TestRulesCacheRefusesAnUnverifiedContainer(t *testing.T) {
	unsigned, err := mapper.RulesContainerToBase64(&model.DecodedRulesContainer{
		MinimumDistinctUserSignatures: 2,
	})
	if err != nil {
		t.Fatalf("encode container: %v", err)
	}

	for _, tc := range []struct {
		name string
		body string
	}{
		{
			// Wire-valid container with NO signatures over it.
			name: "no signatures",
			body: `{"result":{"rulesContainer":"` + unsigned + `","rulesSignatures":[]}}`,
		},
		{
			// Signatures present, produced by a key this client does not trust, so the
			// distinct-signer threshold cannot be met.
			name: "signature from an unconfigured key",
			body: `{"result":{"rulesContainer":"` + unsigned +
				`","rulesSignatures":[{"userId":"attacker","signature":"YmFkLXNpZy1ub3QtZWNkc2E="}]}}`,
		},
	} {
		t.Run(tc.name, func(t *testing.T) {
			srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
				w.Header().Set("Content-Type", "application/json")
				_, _ = w.Write([]byte(tc.body))
			}))
			defer srv.Close()

			client, err := NewClient(srv.URL,
				WithCredentials(APIKeyCredentials("test-key", "abcdef0123456789")),
				WithSuperAdminKeysPEM([]string{rulesCacheTestKeyPEM(t)}),
				WithMinValidSignatures(1),
			)
			if err != nil {
				t.Fatalf("NewClient: %v", err)
			}
			defer func() { _ = client.Close() }()

			container, err := client.RulesCache().Get(context.Background())
			if err == nil {
				t.Fatalf("rules cache accepted an unverified container: %+v", container)
			}
			if container != nil {
				t.Errorf("a container was returned alongside the error: %+v", container)
			}
		})
	}
}

func rulesCacheTestKeyPEM(t *testing.T) string {
	t.Helper()
	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		t.Fatalf("generate key: %v", err)
	}
	der, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
	if err != nil {
		t.Fatalf("marshal key: %v", err)
	}
	return string(pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: der}))
}

File: taurus-protect-sdk-typescript/tests/unit/cache/rules-cache-verification.test.ts
/**
 * Gate: the rules container feeding address signature verification must have its
 * SuperAdmin signatures verified before anything reads the HSM public key out of it.
 *
 * This SDK fetched that container through the generated API and the raw mapper, so it
 * was decoded unverified and `rulesSignatures` was discarded — AddressService then
 * verified addresses against an HSM key nothing had authenticated. Go, Java and Python
 * all fetch through their governance service; only TypeScript did not, and nothing
 * tested the path, which is why it survived.
 *
 * Drives the real client wiring rather than an injected mock, because the defect WAS
 * the wiring: every service-level test passed throughout.
 */

import { generateKeyPairSync } from "crypto";

import { ProtectClient } from "../../../src/client";
import { IntegrityError } from "../../../src/errors";
import { AddressesApi } from "../../../src/internal/openapi/apis/AddressesApi";
import { GovernanceRulesApi } from "../../../src/internal/openapi/apis/GovernanceRulesApi";
import { rulesContainerToBase64 } from "../../../src/mappers/protobuf-rules-container-encode";
import { createEmptyRulesContainer } from "../../../src/models/governance-rules";

function superAdminKeyPem(): string {
  const { publicKey } = generateKeyPairSync("ec", { namedCurve: "prime256v1" });
  return publicKey.export({ type: "spki", format: "pem" }).toString();
}

function createClient(): ProtectClient {
  return ProtectClient.create({
    host: "https://protect.example.com",
    apiKey: "test-api-key-12345",
    apiSecret: "aabbccdd11223344556677889900aabbccddeeff",
    superAdminKeysPem: [superAdminKeyPem()],
    minValidSignatures: 1,
  });
}

function stubAddress(): void {
  jest
    .spyOn(AddressesApi.prototype, "walletServiceGetAddress")
    .mockResolvedValue({
      result: {
        id: "1",
        walletId: "100",
        address: "0xaaa",
        currency: "ETH",
        signature: "c2ln",
      },
    } as never);
}

describe("rules container verification on the address path", () => {
  let client: ProtectClient | undefined;

  afterEach(() => {
    if (client && !client.isClosed) {
      client.close();
    }
    client = undefined;
    jest.restoreAllMocks();
  });

  it("refuses an unsigned rules container instead of trusting its HSM key", async () => {
    // Wire-valid container carrying NO SuperAdmin signatures — the shape an on-path
    // attacker or a compromised endpoint supplies to swap in their own HSM key.
    jest
      .spyOn(GovernanceRulesApi.prototype, "ruleServiceGetRules")
      .mockResolvedValue({
        result: {
          rulesContainer: rulesContainerToBase64(createEmptyRulesContainer()),
          rulesSignatures: [],
        },
      } as never);
    stubAddress();

    client = createClient();

    await expect(client.addresses.get(1)).rejects.toThrow(IntegrityError);
  });

  it("refuses a container whose signatures verify against none of the configured keys", async () => {
    // Signatures present but produced by a key the client does not trust, so the
    // distinct-signer threshold cannot be met.
    jest
      .spyOn(GovernanceRulesApi.prototype, "ruleServiceGetRules")
      .mockResolvedValue({
        result: {
          rulesContainer: rulesContainerToBase64(createEmptyRulesContainer()),
          rulesSignatures: [
            { userId: "attacker", signature: "YmFkLXNpZ25hdHVyZS1ub3QtZWNkc2E=" },
          ],
        },
      } as never);
    stubAddress();

    client = createClient();

    await expect(client.addresses.get(1)).rejects.toThrow(IntegrityError);
  });

  it("reaches the governance service rather than the generated rules API", async () => {
    // The cache must not call ruleServiceGetRules on its own: that path skips
    // verification entirely. It reaches it only THROUGH GovernanceRuleService, which
    // verifies first — so a rejection here proves the verifying path ran.
    const rulesSpy = jest
      .spyOn(GovernanceRulesApi.prototype, "ruleServiceGetRules")
      .mockResolvedValue({
        result: {
          rulesContainer: rulesContainerToBase64(createEmptyRulesContainer()),
          rulesSignatures: [],
        },
      } as never);
    stubAddress();

    client = createClient();

    await expect(client.addresses.get(1)).rejects.toThrow(IntegrityError);
    expect(rulesSpy).toHaveBeenCalled();
  });
});
